batch/batch/front_end/query/query_v1.py

```diff
@@ -27,16 +29,16 @@ def parse_list_batches_query_v1(user: str, q: str, last_batch_id: Optional[int])
         if '=' in t:
             k, v = t.split('=', 1)
             condition = """
-((batches.id) IN
- (SELECT batch_id FROM job_group_attributes
+((job_groups.batch_id, job_groups.job_group_id) IN
+ (SELECT batch_id, job_group_id FROM job_group_attributes
   WHERE `key` = %s AND `value` = %s))
 """
             args = [k, v]
         elif t.startswith('has:'):
             k = t[4:]
             condition = """
-((batches.id) IN
- (SELECT batch_id FROM job_group_attributes
+((job_groups.batch_id, job_groups.job_group_id) IN
+ (SELECT batch_id, job_group_id FROM job_group_attributes
```

Comment:
Here's the original as currently written. However, how can you filter on the batch_id when you're listing batches with an implied job group always = 0?

```
mysql> EXPLAIN WITH base_t AS (
    ->   SELECT batches.*, job_groups.job_group_id,
    ->     job_groups_cancelled.id IS NOT NULL AS cancelled,
    ->     job_groups_n_jobs_in_complete_states.n_completed,
    ->     job_groups_n_jobs_in_complete_states.n_succeeded,
    ->     job_groups_n_jobs_in_complete_states.n_failed,
    ->     job_groups_n_jobs_in_complete_states.n_cancelled
    ->   FROM job_groups
    ->   LEFT JOIN batches ON batches.id = job_groups.batch_id
    ->   LEFT JOIN billing_projects ON batches.billing_project = billing_projects.name
    ->   LEFT JOIN job_groups_n_jobs_in_complete_states
    ->     ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
    ->   LEFT JOIN job_groups_cancelled
    ->     ON job_groups.batch_id = job_groups_cancelled.id AND job_groups.job_group_id = job_groups_cancelled.job_group_id
    ->   STRAIGHT_JOIN billing_project_users ON batches.billing_project = billing_project_users.billing_project
    ->   WHERE (billing_project_users.`user` = "ci" AND billing_project_users.billing_project = batches.billing_project) AND NOT deleted AND job_groups.job_group_id = 0 AND ((job_groups.batch_id, job_groups.job_group_id) IN
    ->  (SELECT batch_id, job_group_id FROM job_group_attributes
    ->   WHERE `key` = "deploy" AND `value` = "1"))
    ->   ORDER BY batch_id DESC
    ->   LIMIT 51
    -> )
    -> SELECT base_t.*, cost_t.cost, cost_t.cost_breakdown
    -> FROM base_t
    -> LEFT JOIN LATERAL (
    ->   SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    ->   FROM (
    ->     SELECT resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    ->     FROM aggregated_job_group_resources_v3
    ->     WHERE base_t.id = aggregated_job_group_resources_v3.batch_id AND base_t.job_group_id = aggregated_job_group_resources_v3.job_group_id
    ->     GROUP BY resource_id
    ->   ) AS usage_t
    ->   LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> ) AS cost_t ON TRUE
    -> ORDER BY id DESC;
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+-------------------------------------------+------+----------+----------------------------------------------+
| id | select_type       | table                                | partitions | type   | possible_keys                                                                                                                                 | key                            | key_len | ref                                       | rows | filtered | Extra                                        |
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+-------------------------------------------+------+----------+----------------------------------------------+
|  1 | PRIMARY           | <derived2>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL                           | NULL    | NULL                                      |   51 |   100.00 | Using filesort; Rematerialize (<derived4>)   |
|  1 | PRIMARY           | <derived4>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL                           | NULL    | NULL                                      |    2 |   100.00 | Using where                                  |
|  4 | DEPENDENT DERIVED | <derived5>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL                           | NULL    | NULL                                      |    2 |   100.00 | NULL                                         |
|  4 | DEPENDENT DERIVED | resources                            | NULL       | eq_ref | resource_id                                                                                                                                   | resource_id                    | 4       | usage_t.resource_id                       |    1 |   100.00 | NULL                                         |
|  5 | DEPENDENT DERIVED | aggregated_job_group_resources_v3    | NULL       | ref    | PRIMARY,resource_id                                                                                                                           | PRIMARY                        | 8       | base_t.id                                 |   21 |    10.00 | Using where; Using temporary                 |
|  2 | DERIVED           | job_group_attributes                 | NULL       | ref    | PRIMARY,job_group_attributes_key_value,job_group_attributes_value,job_group_attributes_batch_id_key_value,job_group_attributes_batch_id_value | job_group_attributes_key_value | 1073    | const,const                               | 4774 |    10.00 | Using where; Using temporary; Using filesort |
|  2 | DERIVED           | batches                              | NULL       | eq_ref | PRIMARY,batches_deleted,batches_billing_project_state                                                                                         | PRIMARY                        | 8       | batch.job_group_attributes.batch_id       |    1 |    50.00 | Using where                                  |
|  2 | DERIVED           | billing_projects                     | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY                        | 302     | batch.batches.billing_project             |    1 |   100.00 | Using index                                  |
|  2 | DERIVED           | job_groups                           | NULL       | eq_ref | PRIMARY,job_groups_state_callback,job_groups_time_created,job_groups_time_completed                                                           | PRIMARY                        | 12      | batch.job_group_attributes.batch_id,const |    1 |   100.00 | Using index                                  |
|  2 | DERIVED           | job_groups_n_jobs_in_complete_states | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY                        | 8       | batch.job_group_attributes.batch_id       |    1 |   100.00 | Using where                                  |
|  2 | DERIVED           | job_groups_cancelled                 | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY                        | 8       | batch.job_group_attributes.batch_id       |    1 |   100.00 | Using where                                  |
|  2 | DERIVED           | billing_project_users                | NULL       | eq_ref | PRIMARY,billing_project_users_billing_project_user_cs                                                                                         | PRIMARY                        | 604     | batch.batches.billing_project,const       |    1 |   100.00 | Using index                                  |
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+-------------------------------------------+------+----------+----------------------------------------------+
12 rows in set, 3 warnings (0.00 sec)
```

Reply:
You're right it's really slow. Let me see if I can rewrite this to be better.

Reply:
This is now very fast!

```
mysql> EXPLAIN WITH base_t AS ( SELECT batches.*, job_groups.job_group_id,
    ->     job_groups_cancelled.id IS NOT NULL AS cancelled,
    ->     job_groups_n_jobs_in_complete_states.n_completed,
    ->     job_groups_n_jobs_in_complete_states.n_succeeded,
    ->     job_groups_n_jobs_in_complete_states.n_failed,
    ->     job_groups_n_jobs_in_complete_states.n_cancelled
    -> FROM job_groups
    -> LEFT JOIN batches ON batches.id = job_groups.batch_id
    -> LEFT JOIN billing_projects ON batches.billing_project = billing_projects.name
    -> LEFT JOIN job_groups_n_jobs_in_complete_states
    ->     ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
    ->   LEFT JOIN job_groups_cancelled
    ->     ON job_groups.batch_id = job_groups_cancelled.id AND job_groups.job_group_id = job_groups_cancelled.job_group_id
atch_id = job_groups.batch_id AND job_group_attributes.job_group_id = job_groups.job_group_id AND `key` = "deploy" AND `value` = "1"
) AS jga ON TRUE
WHERE (billing_project_users.`user` = "ci" AND billing_project_users.billing_project = batches.billing_project) AND NOT deleted AND job_groups.job_group_id = 0
ORDER BY batch_id DESC
LIMIT 51
)
SELECT base_t.*, cost_t.cost, cost_t.cost_breakdown
FROM base_t
LEFT JOIN LATERAL (
  SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
  FROM (
    SELECT resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    FROM aggregated_job_group_resources_v3
    WHERE base_t.id = aggregated_job_group_resources_v3.batch_id AND base_t.job_group_id = aggregated_job_group_resources_v3.job_group_id
    GROUP BY resource_id
  ) AS usage_t
  LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
) AS cost_t ON TRUE
ORDER BY id DESC;    ->   STRAIGHT_JOIN billing_project_users ON batches.billing_project = billing_project_users.billing_project
    -> LEFT JOIN LATERAL (
    ->   SELECT 1
    ->   FROM job_group_attributes
    ->   WHERE job_group_attributes.batch_id = job_groups.batch_id AND job_group_attributes.job_group_id = job_groups.job_group_id AND `key` = "deploy" AND `value` = "1"
    -> ) AS jga ON TRUE
    -> WHERE (billing_project_users.`user` = "ci" AND billing_project_users.billing_project = batches.billing_project) AND NOT deleted AND job_groups.job_group_id = 0
    -> ORDER BY batch_id DESC
    -> LIMIT 51
    -> )
    -> SELECT base_t.*, cost_t.cost, cost_t.cost_breakdown
    -> FROM base_t
    -> LEFT JOIN LATERAL (
    ->   SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    ->   FROM (
    ->     SELECT resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    ->     FROM aggregated_job_group_resources_v3
    ->     WHERE base_t.id = aggregated_job_group_resources_v3.batch_id AND base_t.job_group_id = aggregated_job_group_resources_v3.job_group_id
    ->     GROUP BY resource_id
    ->   ) AS usage_t
    ->   LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> ) AS cost_t ON TRUE
    -> ORDER BY id DESC;
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+-------------+---------+-------------------------------------+------+----------+-----------------------------------------------+
| id | select_type       | table                                | partitions | type   | possible_keys                                                                                                                                 | key         | key_len | ref                                 | rows | filtered | Extra                                         |
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+-------------+---------+-------------------------------------+------+----------+-----------------------------------------------+
|  1 | PRIMARY           | <derived2>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL        | NULL    | NULL                                |   51 |   100.00 | Using filesort; Rematerialize (<derived4>)    |
|  1 | PRIMARY           | <derived4>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL        | NULL    | NULL                                |    2 |   100.00 | Using where                                   |
|  4 | DEPENDENT DERIVED | <derived5>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL        | NULL    | NULL                                |    2 |   100.00 | NULL                                          |
|  4 | DEPENDENT DERIVED | resources                            | NULL       | eq_ref | resource_id                                                                                                                                   | resource_id | 4       | usage_t.resource_id                 |    1 |   100.00 | NULL                                          |
|  5 | DEPENDENT DERIVED | aggregated_job_group_resources_v3    | NULL       | ref    | PRIMARY,resource_id                                                                                                                           | PRIMARY     | 8       | base_t.id                           |   21 |    10.00 | Using where; Using temporary                  |
|  2 | DERIVED           | job_groups                           | NULL       | index  | PRIMARY,job_groups_state_callback,job_groups_time_created,job_groups_time_completed                                                           | PRIMARY     | 12      | NULL                                |  102 |    10.00 | Using where; Backward index scan; Using index |
|  2 | DERIVED           | batches                              | NULL       | eq_ref | PRIMARY,batches_deleted,batches_billing_project_state                                                                                         | PRIMARY     | 8       | batch.job_groups.batch_id           |    1 |    50.00 | Using where                                   |
|  2 | DERIVED           | billing_projects                     | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY     | 302     | batch.batches.billing_project       |    1 |   100.00 | Using index                                   |
|  2 | DERIVED           | job_groups_n_jobs_in_complete_states | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY     | 8       | batch.job_groups.batch_id           |    1 |   100.00 | Using where                                   |
|  2 | DERIVED           | job_group_attributes                 | NULL       | eq_ref | PRIMARY,job_group_attributes_key_value,job_group_attributes_value,job_group_attributes_batch_id_key_value,job_group_attributes_batch_id_value | PRIMARY     | 310     | batch.job_groups.batch_id,const     |    1 |   100.00 | Using where                                   |
|  2 | DERIVED           | job_groups_cancelled                 | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY     | 8       | batch.job_groups.batch_id           |    1 |   100.00 | Using where                                   |
|  2 | DERIVED           | billing_project_users                | NULL       | eq_ref | PRIMARY,billing_project_users_billing_project_user_cs                                                                                         | PRIMARY     | 604     | batch.batches.billing_project,const |    1 |   100.00 | Using index                                   |
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+-------------+---------+-------------------------------------+------+----------+-----------------------------------------------+
12 rows in set, 5 warnings (0.01 sec)
```

Reply:
I think it should be a separate project to rewrite those queries without the job groups additions. Sound good?

Reply:
OK, separate PR or issue is fine. We'll keep the current slow code since it at least is symmetric to what exists now.

Yeah the change from
```
|  2 | DERIVED           | job_group_attributes                 | .... | 4774 | ...
```
to 
```
|  2 | DERIVED           | job_group_attributes                 | ... | 310     | ...
```
Seems like a vast improvement.

Reply:
Hmm. I'm not sure I fully grok your use of LEFT JOIN LATERAL (...) ON TRUE, but if it works that's good.

I think this is a case where I might use a different SQL query for attribute based searches. It seems like the most desirable query is to `FROM job_group_attributes` since that is the table on which the WHERE conditions will eliminate a lot of tables. This is the fastest thing I could find:
```sql
mysql> explain SELECT batches.*, job_groups.job_group_id,
    ->   job_groups_cancelled.id IS NOT NULL AS cancelled,
    ->   job_groups_n_jobs_in_complete_states.n_completed,
    ->   job_groups_n_jobs_in_complete_states.n_succeeded,
    ->   job_groups_n_jobs_in_complete_states.n_failed,
    ->   job_groups_n_jobs_in_complete_states.n_cancelled
    -> FROM job_group_attributes
    -> LEFT JOIN job_groups ON (job_groups.job_group_id = job_group_attributes.job_group_id
    ->                          AND job_groups.batch_id = job_group_attributes.batch_id)
    -> LEFT JOIN batches ON batches.id = job_groups.batch_id
    -> LEFT JOIN billing_projects ON batches.billing_project = billing_projects.name
    -> LEFT JOIN job_groups_n_jobs_in_complete_states
    ->   ON job_groups.batch_id = job_groups_n_jobs_in_complete_states.id AND job_groups.job_group_id = job_groups_n_jobs_in_complete_states.job_group_id
    -> LEFT JOIN job_groups_cancelled
    ->   ON job_groups.batch_id = job_groups_cancelled.id AND job_groups.job_group_id = job_groups_cancelled.job_group_id
    -> LEFT JOIN billing_project_users ON batches.billing_project = billing_project_users.billing_project
    -> LEFT JOIN lateral (
    ->   SELECT COALESCE(SUM(`usage` * rate), 0) AS cost, JSON_OBJECTAGG(resources.resource, COALESCE(`usage` * rate, 0)) AS cost_breakdown
    ->   FROM (
    ->     SELECT resource_id, CAST(COALESCE(SUM(`usage`), 0) AS SIGNED) AS `usage`
    ->     FROM aggregated_job_group_resources_v3
    ->     WHERE aggregated_job_group_resources_v3.batch_id = job_groups.batch_id
    ->       AND aggregated_job_group_resources_v3.job_group_id = job_groups.job_group_id
    ->     GROUP BY resource_id
    ->   ) AS usage_t
    ->   LEFT JOIN resources ON usage_t.resource_id = resources.resource_id
    -> ) AS cost_t ON TRUE
    -> WHERE billing_project_users.`user` = "ci"
    ->   AND billing_project_users.billing_project = batches.billing_project
    ->   AND NOT deleted
    ->   AND job_groups.job_group_id = 0
    ->   AND `key` = "deploy"
    ->   AND `value` = "1"
    -> ORDER BY job_group_attributes.batch_id DESC
    -> LIMIT 51
    -> ;
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+-------------------------------------------+------+----------+-----------------------------------------+
| id | select_type       | table                                | partitions | type   | possible_keys                                                                                                                                 | key                            | key_len | ref                                       | rows | filtered | Extra                                   |
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+-------------------------------------------+------+----------+-----------------------------------------+
|  1 | PRIMARY           | job_group_attributes                 | NULL       | ref    | PRIMARY,job_group_attributes_key_value,job_group_attributes_value,job_group_attributes_batch_id_key_value,job_group_attributes_batch_id_value | job_group_attributes_key_value | 1073    | const,const                               | 4775 |    10.00 | Using where; Backward index scan        |
|  1 | PRIMARY           | batches                              | NULL       | eq_ref | PRIMARY,batches_deleted,batches_billing_project_state                                                                                         | PRIMARY                        | 8       | batch.job_group_attributes.batch_id       |    1 |    50.00 | Using where                             |
|  1 | PRIMARY           | billing_project_users                | NULL       | eq_ref | PRIMARY,billing_project_users_billing_project_user_cs                                                                                         | PRIMARY                        | 604     | batch.batches.billing_project,const       |    1 |   100.00 | Using index                             |
|  1 | PRIMARY           | billing_projects                     | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY                        | 302     | batch.batches.billing_project             |    1 |   100.00 | Using index                             |
|  1 | PRIMARY           | job_groups                           | NULL       | eq_ref | PRIMARY,job_groups_state_callback,job_groups_time_created,job_groups_time_completed                                                           | PRIMARY                        | 12      | batch.job_group_attributes.batch_id,const |    1 |   100.00 | Using index; Rematerialize (<derived2>) |
|  1 | PRIMARY           | job_groups_n_jobs_in_complete_states | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY                        | 8       | batch.job_group_attributes.batch_id       |    1 |   100.00 | Using where                             |
|  1 | PRIMARY           | job_groups_cancelled                 | NULL       | eq_ref | PRIMARY                                                                                                                                       | PRIMARY                        | 8       | batch.job_group_attributes.batch_id       |    1 |   100.00 | Using where                             |
|  1 | PRIMARY           | <derived2>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL                           | NULL    | NULL                                      |    2 |   100.00 | Using where                             |
|  2 | DEPENDENT DERIVED | <derived3>                           | NULL       | ALL    | NULL                                                                                                                                          | NULL                           | NULL    | NULL                                      |    2 |   100.00 | NULL                                    |
|  2 | DEPENDENT DERIVED | resources                            | NULL       | eq_ref | resource_id                                                                                                                                   | resource_id                    | 4       | usage_t.resource_id                       |    1 |   100.00 | NULL                                    |
|  3 | DEPENDENT DERIVED | aggregated_job_group_resources_v3    | NULL       | ref    | PRIMARY,resource_id                                                                                                                           | PRIMARY                        | 8       | batch.job_groups.batch_id                 |   21 |    10.00 | Using where; Using temporary            |
+----+-------------------+--------------------------------------+------------+--------+-----------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------+---------+-------------------------------------------+------+----------+-----------------------------------------+
11 rows in set, 3 warnings (0.01 sec)
```

Reply:
Runs in 0.08s in my tests.